File: leetcode/javascript/dynamicProgramming/minimumFallingPathSum(medium)/minimumFallingPathSum.js
// https://leetcode.com/problems/unique-paths-ii/description/

/*
Учитывая матрицу целых чисел размера n x n, верните минимальную сумму любого падающего пути через матрицу.
Падающий путь начинается с любого элемента в первой строке и выбирает элемент в следующей строке, который находится либо
 непосредственно под ним, либо по диагонали влево/вправо. В частности, следующим элементом с позиции (row, col) будет
 (row + 1, col - 1), (row + 1, col) или (row + 1, col + 1).
*/

/*
Input: matrix = [[2,1,3],[6,5,4],[7,8,9]]
Output: 13
Пояснение: Есть два пути падения с минимальной суммой, как показано.
*/

/**
 * @param {number[][]} matrix
 * @return {number}
 */
var minFallingPathSum = function(matrix) {
    const dp = [matrix[0]];

    for (let i = 1; i < matrix.length; i++) {
        dp.push([]);
        for (let j = 0; j < matrix.length; j++) {
            let prev = Infinity;
            let curr = matrix[i][j] + dp[i - 1][j];
            let next = Infinity;
            if (j > 0) {
                prev = matrix[i][j] + dp[i - 1][j - 1];
            }
            if (j < length - 1) {
                next = matrix[i][j] + dp[i - 1][j + 1];
            }
            dp[i].push(Math.min(prev, curr, next));
        }
    }

    return Math.min(...dp[dp.length - 1]);
};

const matrix = [[2,1,3],[6,5,4],[7,8,9]];
console.log(minFallingPathSum(matrix)); // 13

File: leetcode/go/dynamicProgramming/minimumFallingPathSum(medium)/minimumFallingPathSum.go
package main

import (
	"fmt"
	"math"
)

/*
https://leetcode.com/problems/minimum-falling-path-sum/description/

Дана матрица n x n целых чисел, вернуть минимальную сумму любого нисходящего пути через матрицу.

Нисходящий путь начинается с любого элемента в первой строке и выбирает элемент в следующей строке, который находится
либо непосредственно под ним, либо по диагонали слева/справа. В частности, следующим элементом из позиции (row, col)
будет (row + 1, col - 1), (row + 1, col) или (row + 1, col + 1).

Input: matrix = [[2,1,3],[6,5,4],[7,8,9]]
Output: 13
Пояснение: Существуют два падающих пути с минимальной суммой, как показано на рисунке.
*/

func main() {
	matrix := [][]int{
		{2, 1, 3},
		{6, 5, 4},
		{7, 8, 9},
	}
	fmt.Println(minFallingPathSum(matrix)) // 13 (путь 1 → 5 → 7)
}

// minFallingPathSum вычисляет минимальную сумму падающего пути в матрице
func minFallingPathSum(matrix [][]int) int {
	n := len(matrix)
	if n == 0 {
		return 0
	}

	// Создаем копию матрицы для хранения промежуточных результатов
	dp := make([][]int, n)
	for i := range dp {
		dp[i] = make([]int, n)
		copy(dp[i], matrix[i])
	}

	for i := 1; i < n; i++ {
		for j := 0; j < n; j++ {
			// Инициализируем минимальную сумму как текущую ячейку + значение сверху
			minSum := dp[i-1][j] + matrix[i][j]

			// Проверяем левую верхнюю диагональ (если существует)
			if j > 0 {
				minSum = int(math.Min(float64(minSum), float64(dp[i-1][j-1]+matrix[i][j])))
			}

			// Проверяем правую верхнюю диагональ (если существует)
			if j < n-1 {
				minSum = int(math.Min(float64(minSum), float64(dp[i-1][j+1]+matrix[i][j])))
			}

			dp[i][j] = minSum
		}
	}

	// Находим минимальное значение в последней строке
	minTotal := dp[n-1][0]
	for j := 1; j < n; j++ {
		if dp[n-1][j] < minTotal {
			minTotal = dp[n-1][j]
		}
	}

	return minTotal
}
